### PR TITLE
chore(tests): Scaffold tests from spec files

### PR DESCRIPTION
**Pull request by avodovnik**
This PR makes a minor change to the `ApiChecker` tool in that it enables generating scaffolds for tests based on the `.spec.ts` file.

```PowerShell
.\ApiChecker.exe scaffold-test --specfile tap.specs.ts --outputfile TapTests.cs
```
will generate:

```Csharp
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using PlaywrightSharp.Tests.BaseTests;
using System.Threading.Tasks;
using Xunit;
using Xunit.Abstractions;


namespace PlaywrightSharp.Tests
{
    
    
    /// <playwright-file>tap.specs.ts</playwright-file>
    [Collection(TestConstants.TestFixtureBrowserCollectionName)]
    public sealed class TapTests : PlaywrightSharpBrowserBaseTest
    {
        
        /// <inheritdoc/>
        public TapTests(ITestOutputHelper output) : 
                base(output)
        {
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should send all of the correct events</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldSendAllOfTheCorrectEvents()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should not send mouse events touchstart is canceled</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldNotSendMouseEventsTouchstartIsCanceled()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should not send mouse events when touchend is canceled</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldNotSendMouseEventsWhenTouchendIsCanceled()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should wait for a navigation caused by a tap</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldWaitForANavigationCausedByATap()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should work with modifiers</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldWorkWithModifiers()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should send well formed touch points</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldSendWellFormedTouchPoints()
        {
            throw new System.NotImplementedException();
        }
        
        /// <playwright-file>tap.specs.ts</playwright-file>
        /// <playwright-it>should wait until an element is visible to tap it</playwright-it>
        [Fact(Timeout=PlaywrightSharp.Playwright.DefaultTimeout)]
        public async Task ShouldWaitUntilAnElementIsVisibleToTapIt()
        {
            throw new System.NotImplementedException();
        }
    }
}
```

Also, I was having some problems building the project, so I changed the `csproj` file. I think there was a problem with the `include` part.

### CODE REVIEW

File: src/tools/ApiChecker/ApiChecker.csproj [hunk 1]
```diff
@@ -1,20 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
 <Project Sdk="Microsoft.NET.Sdk">
-
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp3.1</TargetFramework>
-    <ReleaseVersion>0.0.0</ReleaseVersion>
   </PropertyGroup>
-
+  <Import Project="../../Common/PackageInfo.props" />
```
Comment (kblok) [inline src/tools/ApiChecker/ApiChecker.csproj:7]: We could also add `Dependencies.props` so we analyzers.

**Review by kblok — CHANGES_REQUESTED** [review]

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+            targetUnit.Namespaces.Add(globalNamespace);
+
+            var codeNamespace = new CodeNamespace(@namespace);
+            var targetClass = new CodeTypeDeclaration(@class)
+            {
+                IsClass = true,
+                TypeAttributes = System.Reflection.TypeAttributes.Public | System.Reflection.TypeAttributes.Sealed,
+            };
+
+            targetClass.BaseTypes.Add(new CodeTypeReference("PlaywrightSharpBrowserBaseTest"));
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:92]: I think that `PlaywrightSharpPageBaseTest` by default will be better.

**Review by kblok — CHANGES_REQUESTED** [review]

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                Name = name
+            };
+
+            @class.Namespaces[1].Types[0].Members.Add(method);
+
+            method.Comments.Add(new CodeCommentStatement($"<playwright-file>{testOrigin}</playwright-file>", true));
+            method.Comments.Add(new CodeCommentStatement($"<playwright-it>{testDescribe}</playwright-it>", true));
+            method.CustomAttributes.Add(new CodeAttributeDeclaration("Fact",
+            new CodeAttributeArgument[] {
+                new CodeAttributeArgument("Timeout", new CodeFieldReferenceExpression(new CodeTypeReferenceExpression("PlaywrightSharp.Playwright"), "DefaultTimeout"))
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:139]: Maybe create the test with the `Skip` attribute? So we could create tests, push them, and work on those in other PRs.

**Review by kblok — CHANGES_REQUESTED** [review]

File: src/tools/ApiChecker/ApiChecker.csproj (hunk 1 above)
Comment (avodovnik) [reply src/tools/ApiChecker/ApiChecker.csproj:7]: Yeah, good point. It means changing the `Dependencies.props` though. But that's fine, I think.

File: src/tools/ApiChecker/CheckerOptions.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+using CommandLine;
+
+namespace ApiChecker
+{
+    [Verb("check-api", isDefault: true)]
+    internal class CheckerOptions
+    {
+        // this is empty for now
```
Comment (kblok) [inline src/tools/ApiChecker/CheckerOptions.cs:8]:
```suggestion
```

File: src/tools/ApiChecker/Program.cs [hunk 5]
```diff
@@ -1,429 +1,18 @@
...
-                report.AppendLine("</li>");
-            }
+            return Parser.Default.ParseArguments<ScaffoldTestOptions, CheckerOptions>(args).MapResult(
+                async (ScaffoldTestOptions opts) => await DoScaffoldTest(opts).ConfigureAwait(false),
+                async (CheckerOptions opts) => await Checker.Run().ConfigureAwait(false),
+                _ => Task.CompletedTask);
         }
 
-        public static string ToHtml(this string value) => WebUtility.HtmlEncode(value);
+        private static Task DoScaffoldTest(ScaffoldTestOptions opts) => ScaffoldTest.Run(opts);
```
Comment (kblok) [inline src/tools/ApiChecker/Program.cs:16]: All async tasks should have the Async suffix.

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+using System.Threading.Tasks;
+
+namespace ApiChecker
+{
+    [SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1118:ParameterMustNotSpanMultipleLines", Justification = "CodeDom is complicated.")]
+    internal static class ScaffoldTest
+    {
+        private static readonly TextInfo _textInfo = CultureInfo.InvariantCulture.TextInfo;
+
+        public static Task Run(ScaffoldTestOptions options)
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:17]: If there is not async action, shouldn't `ParseArguments` support a `void` method?

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                return Task.FromException(new FileNotFoundException());
+            }
+
+            var fileInfo = new FileInfo(options.SpecFile);
+
+            int dotSeparator = fileInfo.Name.IndexOf('.');
+            string name = _textInfo.ToTitleCase(fileInfo.Name.Substring(0, dotSeparator)) + "Tests";
+            var targetClass = GenerateClass(options.Namespace, name, fileInfo.Name);
+
+            // this is a naive implementation, but it'll do
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:30]:
```suggestion
```

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                AddTest(targetClass, m.Groups[1].Value, fileInfo.Name);
+            }
+
+            using CodeDomProvider provider = CodeDomProvider.CreateProvider("CSharp");
+            CodeGeneratorOptions codegenOptions = new CodeGeneratorOptions()
+            {
+                BracingStyle = "C",
+            };
+
+            using (StreamWriter sourceWriter = new StreamWriter(options.OutputFile))
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:53]: Can't we use a using statement as in line 47?

File: src/tools/ApiChecker/ScaffoldTest.cs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                    targetClass, sourceWriter, codegenOptions);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static CodeCompileUnit GenerateClass(string @namespace, string @class, string fileOrigin)
+        {
+            var targetUnit = new CodeCompileUnit();
+
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTest.cs:65]:
```suggestion
```

File: src/tools/ApiChecker/ScaffoldTestOptions.cs [hunk 10]
```diff
@@ -0,0 +1,22 @@
+using CommandLine;
+
+namespace ApiChecker
+{
+#pragma warning disable SA1600 // Elements are self-documenting and don't need additional docs
```
Comment (kblok) [inline src/tools/ApiChecker/ScaffoldTestOptions.cs:5]: I would add the comments and make the analyzer happy instead.

**Review by kblok — CHANGES_REQUESTED** [review]
If you want to handle the warnings on Checker.cs that would be great. If not just tell me, I'll create an issue for that.

File: src/tools/ApiChecker/CheckerOptions.cs (hunk 4 above)
Comment (avodovnik) [reply src/tools/ApiChecker/CheckerOptions.cs:8]: Let's open an issue and resolve separately. I'm conscious of keeping the PR as minimal as possible :)

File: src/tools/ApiChecker/ScaffoldTestOptions.cs (hunk 10 above)
Comment (avodovnik) [reply src/tools/ApiChecker/ScaffoldTestOptions.cs:5]: As per our conversation, I've marked it `internal`

File: src/tools/ApiChecker/CheckerOptions.cs (hunk 4 above)
Comment (avodovnik) [reply src/tools/ApiChecker/CheckerOptions.cs:8]: https://github.com/microsoft/playwright-sharp/issues/1010

**Review by kblok — APPROVED** [review]

**Pull request merged**